### PR TITLE
[train] update TORCH_NCCL_ASYNC_ERROR_HANDLING env var

### LINKED ISSUE #47286: [train] Update torch NCCL env var from `NCCL_ASYNC_ERROR_HANDLING` to `TORCH_NCCL_ASYNC_ERROR_HANDLING`
```
(RayTrainWorker pid=2273, ip=172.24.21.66) [rank7]:[W Utils.hpp:108] Warning: Environment variable NCCL_ASYNC_ERROR_HANDLING is deprecated; use TORCH_NCCL_ASYNC_ERROR_HANDLING instead (function getCvarString)
```

https://github.com/ray-project/ray/blob/6e3e33c02c4b50741016ebd359c78fab6ecbf807/python/ray/train/torch/config.py#L107

### PR DESCRIPTION
**Pull request by matthewdeng**
## Why are these changes needed?

Fixes this warning:

```
(RayTrainWorker pid=2273, ip=172.24.21.66) [rank7]:[W Utils.hpp:108] Warning: Environment variable NCCL_ASYNC_ERROR_HANDLING is deprecated; use TORCH_NCCL_ASYNC_ERROR_HANDLING instead (function getCvarString)
```

This environment variable was changed in `torch 2.2.0`: https://github.com/pytorch/pytorch/commit/8f8722e3f1d1121345b80428117be89209ccd772.

More details about this environment variable can be found here: https://github.com/pytorch/pytorch/blob/dc8bb2636c5412c68e72dc6e33a6589683d8fd12/docs/source/torch_nccl_environment_variables.rst.

## Related issue number

Closes https://github.com/ray-project/ray/issues/47286

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/torch/config.py [hunk 1]
```diff
@@ -90,21 +91,24 @@ def _setup_torch_process_group(
...
+            TORCH_NCCL_BLOCKING_WAIT_ENV_VAR = "NCCL_BLOCKING_WAIT"
+        else:
+            TORCH_NCCL_ASYNC_ERROR_HANDLING_ENV_VAR = "TORCH_NCCL_ASYNC_ERROR_HANDLING"
+            TORCH_NCCL_BLOCKING_WAIT_ENV_VAR = "TORCH_NCCL_BLOCKING_WAIT"
+        if (
+            TORCH_NCCL_ASYNC_ERROR_HANDLING_ENV_VAR not in os.environ
+            and TORCH_NCCL_BLOCKING_WAIT_ENV_VAR not in os.environ
+        ):
+            logger.debug(
+                f"Setting {TORCH_NCCL_ASYNC_ERROR_HANDLING_ENV_VAR} to fail if NCCL collective communication operations are timing out. "  # noqa: E501
```
Comment (justinvyu) [inline python/ray/train/torch/config.py:108]:
```suggestion
                f"Setting {TORCH_NCCL_ASYNC_ERROR_HANDLING_ENV_VAR}=1 to fail if NCCL collective communication operations are timing out. "  # noqa: E501
```

File: python/ray/train/torch/config.py [hunk 2]
```diff
@@ -90,21 +91,24 @@ def _setup_torch_process_group(
...
-        logger.debug(
-            "Setting NCCL_ASYNC_ERROR_HANDLING to fail if NCCL collective "
-            "communication operations are timing out. "
-            "To override this behavior, you can set NCCL_ASYNC_ERROR_HANDLING=0."
-        )
-        os.environ["NCCL_ASYNC_ERROR_HANDLING"] = "1"
+    if backend == "nccl":
+        # See https://github.com/pytorch/pytorch/blob/c263bd43e8e8502d4726643bc6fd046f0130ac0e/torch/distributed/distributed_c10d.py#L803-L823 # noqa: E501
+        # We do not use TORCH_NCCL_BLOCKING_WAIT due to performance overhead.
+        if Version(torch.__version__) < Version("2.2.0"):
```
Comment (justinvyu) [inline python/ray/train/torch/config.py:97]: how'd you find this version in the end?

**Review by justinvyu — APPROVED** [review]
Thanks!

File: python/ray/train/torch/config.py (hunk 2 above)
Comment (matthewdeng) [reply python/ray/train/torch/config.py:97]: Oops I haven't updated the PR description, but it was added in https://github.com/pytorch/pytorch/commit/8f8722e3f1d1121345b80428117be89209ccd772!

**Pull request merged**